### PR TITLE
[Ballista] Add Decimal128, Date64, TimestampSecond, TimestampMillisecond, Interv…

### PR DESCRIPTION
**Pull request by gaojun2048**
close https://github.com/apache/arrow-datafusion/issues/1658

### CODE REVIEW

**Review by houqp — APPROVED** [review]
LGTM, cc @realno @yahoNanJing

File: ballista/rust/core/proto/ballista.proto [hunk 1]
```diff
@@ -1179,9 +1179,21 @@ message ScalarValue{
...
+        int64 date_64_value = 21;
+        int64 time_second_value = 22;
+        int64 time_millisecond_value = 23;
+        int32 interval_yearmonth_value = 24;
+        int64 interval_daytime_value = 25;
     }
 }
 
+message Decimal128{
+  string value = 1;
```
Comment (liukun4515) [inline ballista/rust/core/proto/ballista.proto:1192]: I'm confused about this point, we use the `String` type to hold the value of Decimal.

**Comment by liukun4515** [comment]
@houqp  I will review this carefully later

File: ballista/rust/core/proto/ballista.proto (hunk 1 above)
Comment (gaojun2048) [reply ballista/rust/core/proto/ballista.proto:1192]: > I'm confused about this point, we use the `String` type to hold the value of Decimal.

The value type is `i128` in datafusion Decimal128, But I don't know how to define i128 in proto. Or is it possible to use int64 in proto?

Comment (houqp) [reply ballista/rust/core/proto/ballista.proto:1192]: Better use raw bytes.

Comment (gaojun2048) [reply ballista/rust/core/proto/ballista.proto:1192]: > Better use 16 raw bytes.

Yes, bytes is better. I will change the code

Comment (gaojun2048) [reply ballista/rust/core/proto/ballista.proto]: done.

**Comment by alamb** [comment]
https://github.com/apache/arrow-datafusion/runs/4919106498?check_suite_focus=true

```
error[E0308]: mismatched types
   --> C:\Users\runneradmin\.cargo\registry\src\github.com-1ecc6299db9ec823\which-4.2.3\src\finder.rs:161:65
    |
161 | ...                   if s.as_bytes().first() == Some(b'.') {
    |                                                       ^^^^
    |                                                       |
    |                                                       expected `&u8`, found `u8`
    |                                                       help: consider borrowing here: `&b'.'`

For more information about this error, try `rustc --explain E0308`.
error: could not compile `which` due to previous error
```

This looks like a bug in the `which` crate which was fixed: https://github.com/harryfei/which-rs/compare/4.2.3...4.2.4

I think it will pass on a rerun -- I retriggered he check and hopefully it will pass this time

**Comment by gaojun2048** [comment]
> https://github.com/apache/arrow-datafusion/runs/4919106498?check_suite_focus=true
> 
> ```
> error[E0308]: mismatched types
>    --> C:\Users\runneradmin\.cargo\registry\src\github.com-1ecc6299db9ec823\which-4.2.3\src\finder.rs:161:65
>     |
> 161 | ...                   if s.as_bytes().first() == Some(b'.') {
>     |                                                       ^^^^
>     |                                                       |
>     |                                                       expected `&u8`, found `u8`
>     |                                                       help: consider borrowing here: `&b'.'`
> 
> For more information about this error, try `rustc --explain E0308`.
> error: could not compile `which` due to previous error
> ```
> 
> This looks like a bug in the `which` crate which was fixed: [harryfei/which-rs@4.2.3...4.2.4](https://github.com/harryfei/which-rs/compare/4.2.3...4.2.4)
> 
> I think it will pass on a rerun -- I retriggered he check and hopefully it will pass this time

Yes. It's ok now.

**Comment by realno** [comment]
LGTM

**Review by realno — APPROVED** [review]

**Comment by alamb** [comment]
Thanks @realno  and @gaojun2048

**Pull request merged**